### ISSUE TITLE
AlexNet Weight Freeze

### ISSUE THREAD

**Issue by snownus** [tensorpack/tensorpack#1204]
### 1. What you did:

I run the DoReFaNet AlexNet example using 1bit activation, 1bit weight and 32-bit gradients. 

I want to freeze the weights before the `logits` layer. 

### 2. What you observed:

[0522 12:53:06 @gradproc.py:77] WRN No gradient w.r.t 20 trainable variables: conv0/W, conv0/b, conv1/W, bn1/gamma, bn1/beta, conv2/W, bn2/gamma, bn2/beta, conv3/W, bn3/gamma, bn3/beta, conv4/W, bn4/gamma, bn4/beta, fc0/b, bnfc0/gamma, bnfc0/beta, bnfc1/gamma, bnfc1/beta, fct/b
[0522 12:53:07 @model_utils.py:64] Trainable Variables: 
name            shape                  dim
--------------  ----------------  --------
conv0/W:0       [12, 12, 3, 96]      41472
conv0/b:0       [96]                    96
conv1/W:0       [5, 5, 48, 256]     307200
bn1/gamma:0     [256]                  256
bn1/beta:0      [256]                  256
conv2/W:0       [3, 3, 256, 384]    884736
bn2/gamma:0     [384]                  384
bn2/beta:0      [384]                  384
conv3/W:0       [3, 3, 192, 384]    663552
bn3/gamma:0     [384]                  384
bn3/beta:0      [384]                  384
conv4/W:0       [3, 3, 192, 256]    442368
bn4/gamma:0     [256]                  256
bn4/beta:0      [256]                  256
fc0/W:0         [9216, 4096]      37748736
fc0/b:0         [4096]                4096
bnfc0/gamma:0   [4096]                4096
bnfc0/beta:0    [4096]                4096
fc1/W:0         [4096, 4096]      16777216
bnfc1/gamma:0   [4096]                4096
bnfc1/beta:0    [4096]                4096
fct/W:0         [4096, 1000]       4096000
fct/b:0         [1000]                1000

### 3. What you expected, if not obvious.

I expect that the weights of fully connected layers are freezed. Unexpectedly, they are not.  The gradients of **fc0/W:0**, **fc1/W:0** and **fct/W:0** still exists. 

### 4. Your environment:
Hardware: 4GPUs Titan Ti.
Python 2.7

@ppwwyyxx , Could you help? Thanks!

**Comment by ppwwyyxx** [comment]
https://tensorpack.readthedocs.io/tutorial/faq.html#how-to-freeze-some-variables-in-training

**Comment by snownus** [comment]
@ppwwyyxx ， yes, I have tried to use the tf.stop_gradient(), but it only freeze the first several conv layers' weights. The fully connected layers' weights are not freezed. Can you explain why?

**Comment by ppwwyyxx** [comment]
If you have done something, please include them in:

### 1. What you did:

(1) **If you're using examples, what's the command you run:**

(2) **If you're using examples, have you made any changes to the examples? Paste `git status; git diff` here:**

(3) **If not using examples, tell us what you did:**

**Comment by ppwwyyxx** [comment]
It seems that you did not modify anything about weight decay, therefore the FC layer's weights are still trained by weight decay. You can change it at :+1: https://github.com/tensorpack/tensorpack/blob/e97ce38dad5af583763f836ccc87ba14ea52f813/examples/DoReFa-Net/alexnet-dorefa.py#L63-L64

**Comment by snownus** [comment]
@ppwwyyxx The gradients look normal now! Thanks for your suggestions! However, after using **tf.stop_gradients on logits** and remove **RunUpdateOps**, the training error still changes after several epochs. In my codes, I only add another fully connected layer after logits, followed by a loss layer. It should not change the training error. Can you provide some suggestions? I thought it would be some tensorpack trick i didn't know. Thanks in advance!

**Comment by ppwwyyxx** [comment]
I could not understand what you did. Please post according to the issue template.

**Comment by snownus** [comment]
@ppwwyyxx , 

**1. What I did:**

- I try to freeze the weights and batch norm parameter by using **tf.stop_gradients on logits and remove RunUpdateOps in callback operation**. 

- I try to add another fully connected layer + loss layer after **logits** layer of AlexNet. 
Here is the code
'
stoped_logits = tf.stop_gradient(logits)
output = FullyConnected('confidence', stoped_logits, 1)
add_param_summary(('.*/W', ['histogram', 'rms']))
outputt = (tf.sign(output) + 1) / 2
conf_correct = prediction_conf_correct(correct, outputt, name='conf-correct-top1')
add_moving_summary(tf.reduce_mean(conf_correct, name='conf-train-correct-top1'))
'

**2. What I observed**
The training error changes per epoch, but validation error doesn't change.

Here is the log
[0522 19:01:42 @monitor.py:459] QueueInput/queue_size: 1.6936e-38
[0522 19:01:42 @monitor.py:459] conf-train-correct-top1: 0.5078
[0522 19:01:42 @monitor.py:459] l2_regularize_loss: 2.263e-37
[0522 19:01:42 @monitor.py:459] learning_rate: 0.01
[0522 19:01:42 @monitor.py:459] param-summary/confidence/W-rms: 0
[0522 19:01:42 @monitor.py:459] param-summary/conv0/W-rms: 0.10412
[0522 19:01:42 @monitor.py:459] param-summary/conv0/W-rms_1: 0.10412
[0522 19:01:42 @monitor.py:459] param-summary/conv1/W-rms: 0.4558
[0522 19:01:42 @monitor.py:459] param-summary/conv1/W-rms_1: 0.4558
[0522 19:01:42 @monitor.py:459] param-summary/conv2/W-rms: 0.31383
[0522 19:01:42 @monitor.py:459] param-summary/conv2/W-rms_1: 0.31383
[0522 19:01:42 @monitor.py:459] param-summary/conv3/W-rms: 0.2774
[0522 19:01:42 @monitor.py:459] param-summary/conv3/W-rms_1: 0.2774
[0522 19:01:42 @monitor.py:459] param-summary/conv4/W-rms: 0.3697
[0522 19:01:42 @monitor.py:459] param-summary/conv4/W-rms_1: 0.3697
[0522 19:01:42 @monitor.py:459] param-summary/fc0/W-rms: 0.042924
[0522 19:01:42 @monitor.py:459] param-summary/fc0/W-rms_1: 0.042924
[0522 19:01:42 @monitor.py:459] param-summary/fc1/W-rms: 0.069286
[0522 19:01:42 @monitor.py:459] param-summary/fc1/W-rms_1: 0.069286
[0522 19:01:42 @monitor.py:459] param-summary/fct/W-rms: 0.12619
[0522 19:01:42 @monitor.py:459] param-summary/fct/W-rms_1: 0.12619
[0522 19:01:42 @monitor.py:459] train-error-top1: 0.5078
[0522 19:01:42 @monitor.py:459] train-error-top5: 0.27603
[0522 19:01:42 @monitor.py:459] val-conf-correct-top1: 0.51536
[0522 19:01:42 @monitor.py:459] val-error-top1: 0.51536
[0522 19:01:42 @monitor.py:459] val-error-top5: 0.27698
[0522 19:01:42 @monitor.py:459] xentropy-loss: 3.0478
[0522 19:01:42 @group.py:48] Callbacks took 56.032 sec in total. InferenceRunner: 55.7 seconds
[0522 19:01:42 @base.py:250] Start Epoch 3 ...
100%|##########|5000/5000[13:55<00:00, 5.98it/s]
[0522 19:15:38 @base.py:260] Epoch 3 (global_step 15000) finished, time:13 minutes 55 seconds.
[0522 19:15:38 @saver.py:77] Model saved to train_log/alexnet-dorefa-conf-1,1,32/model-15000.
100%|##########|782/782[00:55<00:00,14.09it/s]
[0522 19:16:33 @monitor.py:459] QueueInput/queue_size: 1.4608e-38
[0522 19:16:33 @monitor.py:459] conf-train-correct-top1: 0.49676
[0522 19:16:33 @monitor.py:459] l2_regularize_loss: 2.263e-37
[0522 19:16:33 @monitor.py:459] learning_rate: 0.01
[0522 19:16:33 @monitor.py:459] param-summary/confidence/W-rms: 0
[0522 19:16:33 @monitor.py:459] param-summary/conv0/W-rms: 0.10412
[0522 19:16:33 @monitor.py:459] param-summary/conv0/W-rms_1: 0.10412
[0522 19:16:33 @monitor.py:459] param-summary/conv1/W-rms: 0.4558
[0522 19:16:33 @monitor.py:459] param-summary/conv1/W-rms_1: 0.4558
[0522 19:16:33 @monitor.py:459] param-summary/conv2/W-rms: 0.31383
[0522 19:16:33 @monitor.py:459] param-summary/conv2/W-rms_1: 0.31383
[0522 19:16:33 @monitor.py:459] param-summary/conv3/W-rms: 0.2774
[0522 19:16:33 @monitor.py:459] param-summary/conv3/W-rms_1: 0.2774
[0522 19:16:33 @monitor.py:459] param-summary/conv4/W-rms: 0.3697
[0522 19:16:33 @monitor.py:459] param-summary/conv4/W-rms_1: 0.3697
[0522 19:16:33 @monitor.py:459] param-summary/fc0/W-rms: 0.042924
[0522 19:16:33 @monitor.py:459] param-summary/fc0/W-rms_1: 0.042924
[0522 19:16:33 @monitor.py:459] param-summary/fc1/W-rms: 0.069286
[0522 19:16:33 @monitor.py:459] param-summary/fc1/W-rms_1: 0.069286
[0522 19:16:33 @monitor.py:459] param-summary/fct/W-rms: 0.12619
[0522 19:16:33 @monitor.py:459] param-summary/fct/W-rms_1: 0.12619
[0522 19:16:33 @monitor.py:459] train-error-top1: 0.51508
[0522 19:16:33 @monitor.py:459] train-error-top5: 0.29268
[0522 19:16:33 @monitor.py:459] val-conf-correct-top1: 0.5228
[0522 19:16:33 @monitor.py:459] val-error-top1: 0.51536
[0522 19:16:33 @monitor.py:459] val-error-top5: 0.27698
[0522 19:16:33 @monitor.py:459] xentropy-loss: 3.1443
[0522 19:16:33 @group.py:48] Callbacks took 55.863 sec in total. InferenceRunner: 55.5 seconds
[0522 19:16:33 @base.py:250] Start Epoch 4 ...
100%|##########|5000/5000[13:56<00:00, 5.98it/s]
[0522 19:30:30 @base.py:260] Epoch 4 (global_step 20000) finished, time:13 minutes 56 seconds.
[0522 19:30:30 @saver.py:77] Model saved to train_log/alexnet-dorefa-conf-1,1,32/model-20000.
  0%|          |0/782[00:00<?,?it/s][0522 19:31:27 @monitor.py:459] QueueInput/queue_size: 1.7146e-38
[0522 19:31:27 @monitor.py:459] conf-train-correct-top1: 0.51858
[0522 19:31:27 @monitor.py:459] l2_regularize_loss: 2.263e-37
[0522 19:31:27 @monitor.py:459] learning_rate: 0.01
[0522 19:31:27 @monitor.py:459] param-summary/confidence/W-rms: 0
[0522 19:31:27 @monitor.py:459] param-summary/conv0/W-rms: 0.10412
[0522 19:31:27 @monitor.py:459] param-summary/conv0/W-rms_1: 0.10412
[0522 19:31:27 @monitor.py:459] param-summary/conv1/W-rms: 0.4558
[0522 19:31:27 @monitor.py:459] param-summary/conv1/W-rms_1: 0.4558
[0522 19:31:27 @monitor.py:459] param-summary/conv2/W-rms: 0.31383
[0522 19:31:27 @monitor.py:459] param-summary/conv2/W-rms_1: 0.31383
[0522 19:31:27 @monitor.py:459] param-summary/conv3/W-rms: 0.2774
[0522 19:31:27 @monitor.py:459] param-summary/conv3/W-rms_1: 0.2774
[0522 19:31:27 @monitor.py:459] param-summary/conv4/W-rms: 0.3697
[0522 19:31:27 @monitor.py:459] param-summary/conv4/W-rms_1: 0.3697
[0522 19:31:27 @monitor.py:459] param-summary/fc0/W-rms: 0.042924
[0522 19:31:27 @monitor.py:459] param-summary/fc0/W-rms_1: 0.042924
[0522 19:31:27 @monitor.py:459] param-summary/fc1/W-rms: 0.069286
[0522 19:31:27 @monitor.py:459] param-summary/fc1/W-rms_1: 0.069286
[0522 19:31:27 @monitor.py:459] param-summary/fct/W-rms: 0.12619
[0522 19:31:27 @monitor.py:459] param-summary/fct/W-rms_1: 0.12619
[0522 19:31:27 @monitor.py:459] train-error-top1: 0.51022
[0522 19:31:27 @monitor.py:459] train-error-top5: 0.27873
[0522 19:31:27 @monitor.py:459] val-conf-correct-top1: 0.5228
[0522 19:31:27 @monitor.py:459] val-error-top1: 0.51536
[0522 19:31:27 @monitor.py:459] val-error-top5: 0.27698
[0522 19:31:27 @monitor.py:459] xentropy-loss: 3.1212
[0522 19:31:27 @group.py:48] Callbacks took 56.623 sec in total. InferenceRunner: 56.3 seconds

**3. What I expected:**
The training error and validation error don't change. 

The additional fully connected layer's parameter being updated and loss layer I added changed. But in my log, it is not.

Please help, thanks! @ppwwyyxx

**Comment by ppwwyyxx** [comment]
> The training error and validation error don't change.

That is not a correct expectation. Training error depends on what images are sampled for training and what random augmentations they went through. So it will change although the weights are freezed.

**Comment by snownus** [comment]
@ppwwyyxx , yes, you are right. How about the validation error? As I kept the gradients updating on the weights of additional fully connected layer, the validation error should change, right? Sorry for so many disturbing. Your suggestion is really useful for me.
